// coffee-coverage/lib/CompiledCache.js
// Generated by CoffeeScript 1.7.1
(function() {
  var CompiledCache, EXTENSIONS, fs, getRelativeFilename, mkdirs, path, _, _ref;

  fs = require('fs');

  path = require('path');

  _ = require('lodash');

  EXTENSIONS = require('./constants').EXTENSIONS;

  _ref = require('./utils/helpers'), mkdirs = _ref.mkdirs, getRelativeFilename = _ref.getRelativeFilename;

  module.exports = CompiledCache = (function() {
    function CompiledCache(basePath, cacheDir, ext) {
      this.basePath = basePath;
      this.cacheDir = cacheDir;
      this.ext = ext != null ? ext : '_covered';
    }

    CompiledCache.prototype._getCacheFileName = function(fileName, options) {
      var cacheFile, newExt, relativeFile, _ref1;
      if (options == null) {
        options = {};
      }
      newExt = (_ref1 = options.ext) != null ? _ref1 : this.ext;
      relativeFile = getRelativeFilename(this.basePath, fileName);
      cacheFile = path.resolve(this.cacheDir, relativeFile);
      cacheFile += newExt;
      return cacheFile;
    };

    CompiledCache.prototype.get = function(fileName, compileFn) {
      var answer, cacheFileName, cacheStat, fileStat;
      if (compileFn == null) {
        compileFn = null;
      }
      if (!this.cacheDir) {
        return typeof compileFn === "function" ? compileFn() : void 0;
      }
      cacheFileName = this._getCacheFileName(fileName);
      answer = null;
      if (fs.existsSync(cacheFileName)) {
        fileStat = fs.statSync(fileName);
        cacheStat = fs.statSync(cacheFileName);
        if (cacheStat.ctime > fileStat.mtime) {
          answer = fs.readFileSync(cacheFileName, {
            encoding: 'utf8'
          });
        }
      }
      if ((answer == null) && (compileFn != null)) {
        answer = compileFn();
        this.put(fileName, answer);
      }
      return answer;
    };

    CompiledCache.prototype.put = function(fileName, contents, options) {
      var cacheDir, cacheFileName;
      if (options == null) {
        options = {};
      }
      if (!this.cacheDir || !contents) {
        return;
      }
      cacheFileName = this._getCacheFileName(fileName, options);
      cacheDir = path.dirname(cacheFileName);
      mkdirs(cacheDir);
      return fs.writeFileSync(cacheFileName, contents, {
        encoding: 'utf8'
      });
    };

    return CompiledCache;

  })();

}).call(this);
