// coffee-coverage/lib/utils/helpers.js
// Generated by CoffeeScript 1.7.1
(function() {
  var EXTENSIONS, assert, fs, path, statFile, _,
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  assert = require('assert');

  fs = require('fs');

  path = require('path');

  _ = require('lodash');

  EXTENSIONS = require('../constants').EXTENSIONS;

  exports.stripLeadingDotOrSlash = function(pathName) {
    return pathName.replace(/^\//, "").replace(/^\.\//, "");
  };

  exports.statFile = statFile = function(file) {
    if (!fs.existsSync(file)) {
      return null;
    }
    return fs.statSync(file);
  };

  exports.mkdirs = function(dirPath, mode) {
    var currentPath, pathElement, pathElements, stat, _i, _len;
    if (!statFile(dirPath)) {
      pathElements = dirPath.split(path.sep);
      if (_.last(pathElements) === '') {
        pathElements.pop();
      }
      currentPath = "";
      for (_i = 0, _len = pathElements.length; _i < _len; _i++) {
        pathElement = pathElements[_i];
        currentPath += pathElement + path.sep;
        stat = statFile(currentPath);
        if (stat && !stat.isDirectory()) {
          throw new CoverageError("Can't create directory " + currentPath + ": file already exists.");
        }
        if (!stat) {
          fs.mkdirSync(currentPath, mode);
        }
      }
      return true;
    }
    return false;
  };

  exports.getRelativeFilename = function(basePath, fileName) {
    if ((basePath != null) && _.startsWith(fileName, basePath)) {
      fileName = path.relative(basePath, fileName);
    }
    return fileName;
  };

  exports.excludeFile = function(fileName, options) {
    var basePath, component, components, exclude, excludePath, excluded, relativeFilename, resolvedFileName, _i, _j, _k, _len, _len1, _len2, _ref;
    basePath = options.basePath;
    exclude = options.exclude;
    resolvedFileName = path.resolve(fileName);
    assert(resolvedFileName === fileName);
    if (!exclude) {
      return;
    }
    excluded = false;
    if (basePath) {
      relativeFilename = exports.getRelativeFilename(basePath, fileName);
      if (relativeFilename === fileName) {
        excluded = true;
      }
      components = relativeFilename.split(path.sep);
      for (_i = 0, _len = components.length; _i < _len; _i++) {
        component = components[_i];
        if (__indexOf.call(exclude, component) >= 0) {
          excluded = true;
        }
      }
      if (!excluded) {
        for (_j = 0, _len1 = exclude.length; _j < _len1; _j++) {
          excludePath = exclude[_j];
          if (_.startsWith("/" + relativeFilename, excludePath) || _.startsWith(relativeFilename, excludePath)) {
            excluded = true;
          }
        }
      }
    }
    if (!excluded && (_ref = !path.extname(fileName), __indexOf.call(Object.keys(EXTENSIONS), _ref) >= 0)) {
      excluded = true;
    }
    if (!excluded) {
      for (_k = 0, _len2 = exclude.length; _k < _len2; _k++) {
        excludePath = exclude[_k];
        if (_.startsWith(fileName, excludePath)) {
          excluded = true;
        }
      }
    }
    return excluded;
  };

  exports.toQuotedString = function(string) {
    var answer;
    answer = string.replace(/\\/g, '\\\\');
    return '"' + (answer.replace(/"/g, '\\\"')) + '"';
  };

}).call(this);
